### PR TITLE
API: Return scalars for scalar inputs to np.real/imag

### LINKED ISSUE #8386: Inconsistent behaviour of numpy.real and numpy.imag functions
The numpy.real and numpy.imag functions show what is IMHO inconsistent behaviour compared to other numpy functions - when they are given a a single value as parameter (as opposed to array) they return an array instead of a single value. Compare this to the behaviour of numpy.angle, numpy.conj, numpy.sin, etc. functions which return a scalar when given a scalar. Example code:

>>> np.__version__
'1.11.1'
>>> a = 1.0 + 1j
>>> np.real(a)
array(1.0)
>>> np.imag(a)
array(1.0)
>>> np.angle(a)
0.78539816339744828
>>> np.conj(a)
(1-1j)

### PR DESCRIPTION
**Pull request by gfyoung**
Returns scalars for scalar inputs for the functions `np.real` and `np.imag`, as consistent with functions like `np.angle` and `np.conj`.

Closes #8386.

### CODE REVIEW

**Comment by seberg** [comment]
Its not obvious that this is correct. doing this implies that putting in a 0-D array will not give you a view like the attribute does.

**Comment by gfyoung** [comment]
@seberg : True, but it is consistent with what these other `numpy` functions do.  It would be preferable to choose one (return `view`) or the other (return scalar).

**Comment by seberg** [comment]
The other functions all return copies though.

**Comment by gfyoung** [comment]
Then in that case we should return scalars across the board, unless there is an advantageous reason why we should return a view for the `0-D` array case that would override consistency.

**Comment by seberg** [comment]
It is a view for all other dimensions, I am just saying that it is an API change, and that I think the comparison to `angle` or `conj` is only half valid.

**Comment by gfyoung** [comment]
@seberg : Fair enough, though I did acknowledge that this is an API change if you read the title of this PR.

**Comment by charris** [comment]
Having a view can be useful as the view can be used as the value of an `out` parameter.

**Comment by gfyoung** [comment]
@charris: I still return 0-D arrays when one is passed in, but when a scalar is passed in, I return scalar.  That seems consistent with what we do with other functions.e

**Comment by seberg** [comment]
@gfyoung you do? In that case, can you make it clear with a test pointing that out? 0-D arrays are rare, but still....

**Comment by gfyoung** [comment]
@seberg , @charris : Added test cases to confirm that 0-D array input gets 0-D array output.  Ready to merge if there are no other concerns.

**Comment by seberg** [comment]
Nope, sorry, but me asking for a test was a trick question really.
Because you do not preserve 0-D arrays to return 0-D arrays (or views)
as it is. Not saying I know this can't be OK anyway, but I have my
doubts.

**Comment by gfyoung** [comment]
> Nope, sorry, but me asking for a test was a trick question really.

What type of comment is this?  Not really sure why you're trying to be "clever".

Also, you should look at the PR because making comments like this.  I wrote the test case out, and the behavior does return 0-D array upon 0-D input as requested.

**Comment by seberg** [comment]
Was a bit evil, but anyway... You did not test for array return, because scalars are equal to arrays in those tests.

**Comment by gfyoung** [comment]
> You did not test for array return, because scalars are equal to arrays in those tests.

Now that's a more useful comment: straightforward and to the point.  I've updated the tests to check.

Please refrain from your "trickery" in the future.

**Comment by gfyoung** [comment]
@charris , @seberg : Updated test to confirm that 0-D array is returned when we pass in 0-D array, and everything is still passing.

File: numpy/lib/type_check.py
Comment (seberg) [inline numpy/lib/type_check.py]: Now it never returns scalars.

Comment (gfyoung) [inline numpy/lib/type_check.py]: I strongly beg to differ on this point.  I just confirmed it myself.

Comment (seberg) [inline numpy/lib/type_check.py]: oops, misread.

Comment (gfyoung) [inline numpy/lib/type_check.py]: Ah, okay.  No worries. :smile:

**Comment by gfyoung** [comment]
@charris , @seberg : Any updates on this?

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Given that we imported `_nx` on the line above, and much of this file uses it already, is there any real point importing `isscalar` here?

File: numpy/lib/tests/test_type_check.py
Comment (eric-wieser) [inline numpy/lib/tests/test_type_check.py]: This doesn't actually check that the scalar is returned - all of these tests would have passed before this patch, I think.

**Review by eric-wieser — CHANGES_REQUESTED** [review]

File: numpy/lib/type_check.py
Comment (gfyoung) [inline numpy/lib/type_check.py]: Fair enough.  Done.

File: numpy/lib/tests/test_type_check.py
Comment (gfyoung) [inline numpy/lib/tests/test_type_check.py]: Ah, that is true.  Done.

**Comment by gfyoung** [comment]
@eric-wieser : Made the requested changes.  Please have a look.

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Nit-pick: _"component of **the** ..."_

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: (same here)

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Not really `float`: `complex128` -> `float64`, `complex32` -> `float32`. But that was here before, s out of scope for this PR, I think

**Review by eric-wieser — APPROVED** [review]
Generally looks good, just some minor nitpicks on doc comments - not sure if this needs a manual release note

File: numpy/lib/type_check.py
Comment (gfyoung) [inline numpy/lib/type_check.py]: Actually, the exact `float` type is platform-dependent.  On my machine, it actually returns `float128` when I pass in `complex128`.

File: numpy/lib/type_check.py
Comment (gfyoung) [inline numpy/lib/type_check.py]: Done.

File: numpy/lib/type_check.py
Comment (gfyoung) [inline numpy/lib/type_check.py]: Done.

**Comment by gfyoung** [comment]
> not sure if this needs a manual release note

IMO I don't believe so.  This change is relatively minor compared to those that would find themselves in the release notes.

File: numpy/lib/type_check.py
Comment (seberg) [inline numpy/lib/type_check.py]: That sounds like a bug if its true?

Comment (gfyoung) [inline numpy/lib/type_check.py]: Can't tell at this point, but I am a little surprised that `complex128` would return `float64` if the architecture is capable of supporting `complex128`.

Comment (eric-wieser) [inline numpy/lib/type_check.py]: > I am a little surprised that complex128 would return float64

Why? 64 bits of real + 64 bits of imaginary = 128 bits of complex

Comment (gfyoung) [inline numpy/lib/type_check.py]: Okay...but then why would I get `float128` returned on my machine then?

Comment (seberg) [inline numpy/lib/type_check.py]: Since the result is a view (unless it happens to be scalar), it frankly seems pretty improbable to me at least for the array case. Can you double check? And if, what system exactly?

Comment (gfyoung) [inline numpy/lib/type_check.py]: Ah, okay.  I should have clarified.  For the scalar case, I get `float128`, but for the array case, I do indeed get `float64`.

Comment (eric-wieser) [inline numpy/lib/type_check.py]: By "the scalar case", you presumably mean `complex` not `np.complex128`

Comment (gfyoung) [inline numpy/lib/type_check.py]: @eric-wieser : That is correct.

**Comment by gfyoung** [comment]
@eric-wieser @seberg : Changes made as requested, and all is green.

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Just thinking about duct typing here... How about:

```python
try:
    return val.real
except AttributeError:
    return asanyarray(val).real
```

Which has the nice bonus that builtin `complex` returns builtin `float` not `np.float64`

**Comment by eric-wieser** [comment]
Tests look good now, but I'm kinda tempted to duck-type this in the same way we duck-type `np.min`

File: numpy/lib/type_check.py
Comment (gfyoung) [inline numpy/lib/type_check.py]: Good point.  Done.

**Comment by gfyoung** [comment]
@eric-wieser @seberg : Changes made as requested, and all is green.

**Comment by eric-wieser** [comment]
Not green for me, but looks good. @seberg, do you agree with the duck-typing I suggested above that @gfyoung has now implemented?

**Comment by gfyoung** [comment]
@seberg : Any updates on this?  I think @eric-wieser is waiting for your response before merging.

**Comment by gfyoung** [comment]
@eric-wieser : @seberg seems to not be responding (either because the message wasn't seen or this is tacit approval).  Ideas of how to move forward?

**Comment by seberg** [comment]
It looks good to me now, sorry if I am slow but don't have much brain cycles for numpy right now, and frankly the first implementations looked too tricky to me to just approve off. Even here I would not mind mentioning it in the release notes, but probably it is not necessary.

File: numpy/lib/type_check.py
Comment (eric-wieser) [inline numpy/lib/type_check.py]: Although now this is inconsistent with `np.conj`, which does not return a native `complex`...

What was the previous implementation?

**Comment by eric-wieser** [comment]
Ugh, I've just realized that this fails to meet the goal of _"Returns scalars for scalar inputs as consistent with functions like np.angle and np.conj"_.

**Comment by seberg** [comment]
Well, this is subclassing hell.... `np.angle` I don't understand why that is a goal, there is no `angle` method after all. `np.conjugate` could indeed be set up to go via the method in principle, since it is pretty much a well defined method within python. This "function that is basically supposed to be an operator" thing is really annoying.

Overall, I think the approach of try/except is consistent with what we do on most methods at least, so I don't mind the change, and doubt right now it creates much trouble. @mhvk you are a subclass guy, maybe you can give your opinion/final decision on whether this is an improvement or not?

**Comment by eric-wieser** [comment]
@gyoung: Fear not, we can fall back on what you had before (below) if this turns out to be a bad idea:

```
arr = asanyarray(val)
return arr.real.item() if isscalar(val) else arr.real
```

**Comment by seberg** [comment]
Not sure I like that, because it does not return views. Though, I admit, if it creates problems in code, it should at least not be silent problems, since the scalars are immutable.

**Comment by eric-wieser** [comment]
@seberg: That comment got edited way too much, you should probably check it still says what you were replying to

**Comment by seberg** [comment]
Yeah, well... we just have to be careful we don't simply replace one problem with another.

**Comment by eric-wieser** [comment]
I agree. I think the above code now does return a view, if passed a viewable object?

**Comment by mhvk** [comment]
I had a look and I think this is the right implementation. I don't think it makes all that much sense to compare with `conj` and `angle`, as both actually do an operation, while `real` and `imag` just select a particular part of an operand. (Indeed, `angle` could use some work - why, o why, this need to `asarray` all input...).

So, I think this should just go in.

**Comment by seberg** [comment]
Isn't it a great convenience ;). Ok, then lets just put it in and hopefully we are not missing anything that might get broken here.

**Comment by charris** [comment]
Needs a release note...

**Comment by gfyoung** [comment]
@seberg , @eric-wieser , @mhvk : Thanks for all of the input!

@charris : Acknowledged.  Will put up PR as follow-up.

**Comment by WarrenWeckesser** [comment]
Take a look at https://github.com/scipy/scipy/issues/7227

Note, in particular, the quirk of `numpy.testing.assert_equal()` that I commented on there (using numpy 1.12.1):

```
In [10]: from numpy.testing import assert_equal

In [11]: assert_equal(np.array(-0.0), 0)

In [12]: assert_equal(-0.0, 0)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-12-8315da2bce84> in <module>()
----> 1 assert_equal(-0.0, 0)

/Users/warren/miniconda3scipy/lib/python3.5/site-packages/numpy/testing/utils.py in assert_equal(actual, desired, err_msg, verbose)
    410         elif desired == 0 and actual == 0:
    411             if not signbit(desired) == signbit(actual):
--> 412                 raise AssertionError(msg)
    413     # If TypeError or ValueError raised while using isnan and co, just handle
    414     # as before

AssertionError: 
Items are not equal:
 ACTUAL: -0.0
 DESIRED: 0
```

I haven't tracked down the ultimiate source of the failure in scipy, but I suspect a use of `assert_equal` that previously passed is now failing because something that was an array containing -0.0 is now the scalar -0.0.

**Comment by WarrenWeckesser** [comment]
I've narrowed down the problem to the issue reported in https://github.com/numpy/numpy/issues/8848, so further discussion should be over there.

**Comment by gfyoung** [comment]
@WarrenWeckesser : Still, yikes...what you indicated above is a bug it seems.

**Pull request merged**
